### PR TITLE
Refresh tokens improvements

### PR DESCRIPTION
**Pull request by s2lomon**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

It's a fix of two potential bugs 

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

It improves handling of oauth2 refresh tokens

> How would you describe this change to a non-technical end user or system administrator?

It makes UI experience smother and prevents us from rejecting possibly correct TokenPairs 
## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java [hunk 1]
```diff
@@ -135,9 +135,12 @@ public String serialize(TokenPair tokenPair)
                 .setIssuer(issuer)
                 .claim(ACCESS_TOKEN_KEY, tokenPair.getAccessToken())
                 .claim(EXPIRATION_TIME_KEY, tokenPair.getExpiration())
-                .claim(REFRESH_TOKEN_KEY, tokenPair.getRefreshToken().orElseThrow(JweTokenSerializer::throwExceptionForNonExistingRefreshToken))
                 .compressWith(COMPRESSION_CODEC);
 
+        if (tokenPair.getRefreshToken().isPresent()) {
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java:140]: Should we issue some kind of warning that the `RefreshToken` is not issued ?

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 2]
```diff
@@ -140,30 +147,37 @@ private Optional<Map<String, Object>> getAccessTokenClaims(TokenPair tokenPair)
...
     private void needAuthentication(ContainerRequestContext request, Optional<TokenPair> tokenPair)
     {
-        // send 401 to REST api calls and redirect to others
-        if (request.getUriInfo().getRequestUri().getPath().startsWith("/ui/api/")) {
-            sendWwwAuthenticate(request, "Unauthorized", ImmutableSet.of(TRINO_FORM_LOGIN));
-            return;
-        }
-        Optional<String> refreshToken = tokenPair.flatMap(TokenPair::getRefreshToken);
+        Optional<String> refreshToken = tokenPair
+                .flatMap(TokenPair::getRefreshToken);
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:151]: nit: This can go in a separate commit

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 3]
```diff
@@ -140,30 +147,37 @@ private Optional<Map<String, Object>> getAccessTokenClaims(TokenPair tokenPair)
...
+        // send 401 to REST api calls and redirect to others
+        if (request.getUriInfo().getRequestUri().getPath().startsWith("/ui/api/")) {
+            sendWwwAuthenticate(request, "Unauthorized", ImmutableSet.of(TRINO_FORM_LOGIN));
         }
         else {
             startOAuth2Challenge(request);
         }
     }
 
+    private void redirectForNewToken(ContainerRequestContext request, String refreshToken)
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:171]: nit: This looks like cleanup, can we move it to a different commit ?

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 4]
```diff
@@ -163,6 +158,10 @@ private void needAuthentication(ContainerRequestContext request, Optional<TokenP
             }
             return;
         }
+        // send 401 to REST api calls and redirect to others
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:161]: This should be next to `else` block right ?

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 5]
```diff
@@ -94,9 +95,10 @@ public void filter(ContainerRequestContext request)
             request.abortWith(Response.seeOther(DISABLED_LOCATION_URI).build());
             return;
         }
-        Optional<Map<String, Object>> claims;
         Optional<TokenPair> tokenPair = getTokenPair(request);
-        claims = tokenPair.flatMap(this::getAccessTokenClaims);
+        Optional<Map<String, Object>> claims = tokenPair
+                .filter(this::tokenNotExpired)
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:100]: Will it ever happen? IIRC JWT library throws an exception when trying to parse an expired token.

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 6]
```diff
@@ -152,19 +152,18 @@ private void needAuthentication(ContainerRequestContext request, Optional<TokenP
             try {
                 redirectForNewToken(request, refreshToken.get());
             }
-            catch (ChallengeFailedException e) {
+            catch (Exception e) {
                 LOG.debug(e, "Tokens refresh challenge has failed");
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:156]: Not really related but since you're already making changes in this area we could consider adding additional tokens logging, probably behind flag, to make debugging easier. Similar to Hydra's `LOG_LEAK_SENSITIVE_VALUES` https://www.ory.sh/docs/ecosystem/logging

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java [hunk 7]
```diff
@@ -22,6 +22,7 @@
 import com.nimbusds.jose.Payload;
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java:22]: Typo in cmt msg

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java (hunk 5 above)
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:100]: This checks is for the access token not for the token we mint. Additionally it reduces the load on the IdP in case of expired access token.

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 8]
```diff
@@ -148,22 +155,29 @@ private void needAuthentication(ContainerRequestContext request, Optional<TokenP
...
-                        .cookie(OAuthWebUiCookie.create(serializedToken, tokenExpiration.map(expiration -> Instant.now().plus(expiration)).orElse(response.getExpiration())))
-                        .build());
+                redirectForNewToken(request, refreshToken.get());
             }
             catch (ChallengeFailedException e) {
                 LOG.debug(e, "Tokens refresh challenge has failed");
-                startOAuth2Challenge(request);
+                sendWwwAuthenticate(request, "Refresh token flow failed", ImmutableSet.of(TRINO_FORM_LOGIN));
             }
+            return;
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:159]: no, because this method is a void one (as we modify the response in Filter, rather than creating and returning it). So I need to return somewhere, to stop further processing. Still I will change it to if-else as it seems clearer than this return;

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java (hunk 7 above)
Comment (s2lomon) [reply core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java:22]: should be fixed.

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 9]
```diff
@@ -147,11 +147,6 @@ private Optional<Map<String, Object>> getAccessTokenClaims(TokenPair tokenPair)
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:147]: done

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 10]
```diff
@@ -152,19 +152,18 @@ private void needAuthentication(ContainerRequestContext request, Optional<TokenP
             try {
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:152]: done.

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java [hunk 11]
```diff
@@ -148,22 +155,29 @@ private void needAuthentication(ContainerRequestContext request, Optional<TokenP
...
-                String serializedToken = tokenPairSerializer.serialize(TokenPair.fromOAuth2Response(response));
-                request.abortWith(Response.seeOther(request.getUriInfo().getRequestUri())
-                        .cookie(OAuthWebUiCookie.create(serializedToken, tokenExpiration.map(expiration -> Instant.now().plus(expiration)).orElse(response.getExpiration())))
-                        .build());
+                redirectForNewToken(request, refreshToken.get());
             }
             catch (ChallengeFailedException e) {
                 LOG.debug(e, "Tokens refresh challenge has failed");
-                startOAuth2Challenge(request);
+                sendWwwAuthenticate(request, "Refresh token flow failed", ImmutableSet.of(TRINO_FORM_LOGIN));
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java]: I think it belongs here, as before we were mixing up handling error from refreshToken flow for api and non api calls.

File: core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java (hunk 6 above)
Comment (11xor6) [reply core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java:157]: I agree with @s2lomon we shouldn't be logging any security tokens, passwords, or other sensitive information. Image that the logs can always be read by a malicious party and act accordingly.

**Review by 11xor6 — APPROVED** [review]
LGTM

**Review by lukasz-walkiewicz — APPROVED** [review]
LGTM 👍

**Comment by Praveen2112** [comment]
CI hit - reported as https://github.com/trinodb/trino/issues/13556, https://github.com/trinodb/trino/issues/13557, https://github.com/trinodb/trino/issues/13507

**Comment by Praveen2112** [comment]
@s2lomon Can you please update the RN here.

**Comment by Praveen2112** [comment]
Thanks for fixing this !!

**Comment by colebow** [comment]
@s2lomon pinging again on the topic of a release note - if you'd like to propose one, I think a bugfix like this would be worth mentioning, especially if any users were running into these bugs.

**Comment by s2lomon** [comment]
> @s2lomon pinging again on the topic of a release note - if you'd like to propose one, I think a bugfix like this would be worth mentioning, especially if any users were running into these bugs.

Sure sorry for that. Take a look at my propositions, as it's fixing few bugs that I've found.
